src/pudl/transform/eia861.py

```diff
@@ -214,47 +324,495 @@ def balancing_authority(raw_dfs, tfr_dfs):
+
+    ###########################################################################
+    # Transform Values:
+    # * Turn 1000s of dollars back into dollars
+    # * Replace report_year (int) with report_date (datetime64[ns])
+    # * Re-code data_observed to boolean:
+    #   * O="observed" => True
+    #   * I="imputed" => False
+    # * Change the form code (A, B, C, D) into the business model that it
+    #   corresponds to (retail vs. energy_services), which in combination with
+    #   the service_type column (energy, delivery, bundled) will now serve as
+    #   part of the primary key for the table.
+    ###########################################################################
+    logger.info("Performing value transformations on EIA 861 Sales table.")
+    transformed_sales = (
+        type_compat_sales.assign(
+            revenues=lambda x: x.revenues * 1000.0,
+            data_observed=lambda x: x.data_observed.replace({
+                "O": True,
+                "I": False,
+            }),
+            business_model=lambda x: x.business_model.replace({
+                "A": "retail",
+                "B": "retail",
+                "C": "retail",
+                "D": "energy_services",
+            }),
+            service_type=lambda x: x.service_type.str.lower(),
+        )
+        .pipe(pudl.helpers.convert_to_date)
+    )
+
+    # REMOVE: when EIA 861 has been integrated with ETL -- this step
+    # should be happening after all of the tables are transformed.
+    # transformed_sales = pudl.helpers.convert_cols_dtypes(
+    #    transformed_sales, "eia", "sales_eia861")
+
+    tfr_dfs["sales_eia861"] = transformed_sales
+
+    return tfr_dfs
+
+
+def advanced_metering_infrastructure(raw_dfs, tfr_dfs):
+    """
+    Transform the EIA 861 Advanced Metering Infrastructure table.
+
+    Args:
+        raw_dfs (dict): A dictionary of raw EIA 861 dataframes, from the extract step.
+        tfr_dfs (dict): A dictionary of transformed EIA 861 DataFrames, keyed by table
+            name. It will be mutated by this function.
+
+    Returns:
+        dict: A dictionary of transformed EIA 861 dataframes, keyed by table name.
+
+    """
+    return tfr_dfs
+
+
+def demand_response(raw_dfs, tfr_dfs):
+    """
+    Transform the EIA 861 Demand Response table.
+
+    Args:
+        raw_dfs (dict): A dictionary of raw EIA 861 dataframes, from the
+        extract step. tfr_dfs (dict): A dictionary of transformed EIA 861
+        DataFrames, keyed by table name. It will be mutated by this function.
+
+    Returns:
+        dict: A dictionary of transformed EIA 861 dataframes, keyed by table
+            name.
+    """
+    idx_cols = [
+        "utility_id_eia",
+        "state",
+        "report_year",
+        "balancing_authority_code_eia",
+    ]
+
+    customer_classes = [
+        "commercial",
+        "industrial",
+        "other",
+        "residential",
+        "total",
+        "transportation"
+    ]
```

Comment:
We should probably make this a module-level constant CUSTOMER_CLASSES since it's going to keep showing up.

Reply:
Great! Can do.